types/ziggy-js/index.d.ts

```diff
@@ -0,0 +1,48 @@
+// Type definitions for ziggy-js 0.9
+// Project: https://github.com/tightenco/ziggy#readme
+// Definitions by: Ben Allfree <https://github.com/benallfree>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// Minimum TypeScript Version: 3.6
+
+export type Param = string | number | string[] | number[] | undefined;
+
+export interface Params {
+    [key: string]: Param;
+}
```

Comment:
If I'm understanding this correctly, `Params` here would be `event` in `route('events.show', event)`, and `Param` would be one of the values inside the `event` object?

If that's the case, the types should be slightly different: `event` (`interface Params`) can be a string, a number, an array (of strings, numbers, or objects), or an object. If it's an object, each value of the object (`type Param`) can be a string, number, object, null, or undefined.

Reply:
I agree with most of this except I don't think an array of objects is valid because `encodeURIComponent` will properly encode an array but not an object.

However, the Typescript definition for `encodeURIComponent` is as follows:

`declare function encodeURIComponent(uriComponent: string | number | boolean): string;`

Since `hydrateUrl` reduces to `encodeURIComponent` for each value, I think the "Typescript" way would be to limit the inputs to compatible types as defined in the official Typescript binding found in `lib.es5.d.ts`.

I've updated it to match.

Reply:
Ziggy inspects the object to find the parameter value and pulls it out, so `encodeURIComponent()` is never passed an object directly. Something like `route('teams.users.show', [2, { name: 'Taylor', id: 12 }])` is allowed, and it'll eventually result in `encodeURIComponent(12)` for that second parameter. Does that make sense?

See https://github.com/tighten/ziggy/blob/0.9.x/tests/js/test.route.js#L223-L231 and https://github.com/tighten/ziggy/blob/0.9.x/src/js/route.js#L71-L103.

Reply:
Hi thanks for those links. I'm sorry if I'm being slow on this, but looking at the `route.js` source it appears that an object will only work as a param value when the object contains a key named `id` https://github.com/tighten/ziggy/blob/0.9.x/src/js/route.js#L99, otherwise it is assumed to be a plain value that can be accepted by `encodeURIComponent`. Do you agree?

I've taken another shot at defining the Params structure accurately, please have another look :)

Reply:
Yep that's right! The way you have it now looks good to me.

This is going to change very soon in `1.0` to allow any key, not just `id`, sorry 🙃 I'll PR a change here to update and will probably tag you for your help looking at it if that's okay.